### PR TITLE
feat: Introduce provisioning of Logs Table for new and existing users

### LINKED ISSUE #641: Introduce provisioning of logs for new indexers
- [x] Uncomment provisioning of logs
- [x] Unit tests in Provisioner
- [x] Integration test

### PR DESCRIPTION
**Pull request by Kevin101Zhang**
First few commits is from this branch https://github.com/near/queryapi/pull/640. Created this branch based off the initial branch.

This PR intends to introduce the creation of the logs table by provisioning the logsSchema and the follow CRON jobs for new Users but does not use or writeLogs to the new logsTable itself.

If this is merged by itself new users will have unused log table but the provisioning will occur.
To provision existing users - https://github.com/near/queryapi/pull/636

### CODE REVIEW

File: runner/src/indexer/indexer.ts [hunk 1]
```diff
@@ -47,6 +47,8 @@ interface IndexerFunction {
 interface Config {
   hasuraAdminSecret: string
   hasuraEndpoint: string
+  hasuraHostOverride?: string
```
Comment (Kevin101Zhang) [inline runner/src/indexer/indexer.ts:50]: Added optional Params for integration test to point integration test indexer to correct host & port

File: runner/src/indexer/indexer.ts [hunk 2]
```diff
@@ -172,12 +181,23 @@ export default class Indexer {
...
         throw e;
       } finally {
-        await Promise.all([...simultaneousPromises]);
+        await Promise.all([...simultaneousPromises/** , (this.deps.indexerMeta as IndexerMeta).writeLogs(logEntries) */]);
       }
     }
     return allMutations;
   }
 
+  // async getDatabaseConnectionParams(hasuraRoleName: string): Promise<DatabaseConnectionParameters>  {
```
Comment (Kevin101Zhang) [inline runner/src/indexer/indexer.ts:190]: Discussed with @morgsmccauley a method used for integration test to point to the proper port and host.

**Comment by Kevin101Zhang** [comment]
PR introduces provisioning for new users. With Morgans PR #636 that has also been merged from main into this PR and uncommented out we also provision existing user's log table conditionally.

1. PR mainly includes uncommenting out provisioning.ts code and removing the skip from provisioning test.
2. Added tested (locally and unit/integration) commented out code in Indexer.ts
3. Updating unit and integration test (commented out)

File: runner/src/indexer/indexer.test.ts [hunk 3]
```diff
@@ -955,8 +965,9 @@ CREATE TABLE
       getDatabaseConnectionParameters: jest.fn().mockReturnValue(genericDbCredentials),
       fetchUserApiProvisioningStatus: jest.fn().mockReturnValue(true),
       provisionUserApi: jest.fn(),
+      provisionLogsIfNeeded: jest.fn(),
```
Comment (Kevin101Zhang) [inline runner/src/indexer/indexer.test.ts:968]: @morgsmccauley I added a few mocks in the test cases for the #636 pass test.

File: runner/src/indexer/indexer.ts [hunk 4]
```diff
@@ -109,33 +111,39 @@ export default class Indexer {
...
-            // logEntries.push({ blockHeight, logTimestamp: new Date(), logType: LogType.SYSTEM, logLevel: LogLevel.INFO, message: `Provisioning endpoint: failure ${error.message}` });
+            // const provisionFailureLogEntry = LogEntry.systemError('Provisioning endpoint: starting', blockHeight);
+            // logEntries.push(provisionFailureLogEntry);
             throw error;
           }
         }
 
-        // logEntries.push({ blockHeight, logTimestamp: new Date(), logType: LogType.SYSTEM, logLevel: LogLevel.INFO, message: runningMessage });
+        const runningLogEntry = LogEntry.systemInfo(runningMessage, blockHeight);
+        logEntries.push(runningLogEntry);
```
Comment (darunrs) [inline runner/src/indexer/indexer.ts:134]: You forgot to comment this out.

File: runner/src/indexer/indexer.ts [hunk 5]
```diff
@@ -109,33 +111,39 @@ export default class Indexer {
...
         // Cache database credentials after provisioning
         const credentialsFetchSpan = this.tracer.startSpan('fetch database connection parameters');
         try {
           this.database_connection_parameters ??= await this.deps.provisioner.getDatabaseConnectionParameters(hasuraRoleName) as DatabaseConnectionParameters;
-          // this.indexer_logger ??= new IndexerLogger(functionName, this.indexer_behavior.log_level, this.database_connection_parameters);
+          // this.database_connection_parameters = await this.getDatabaseConnectionParams(hasuraRoleName);
```
Comment (darunrs) [inline runner/src/indexer/indexer.ts:139]: Why is there a get db conn params function owned by Indexer in addition to the one in provisioner?

File: runner/src/indexer/indexer.ts (hunk 2 above)
Comment (darunrs) [reply runner/src/indexer/indexer.ts:190]: Huh is this like a stop gap solution? I'm wondering what the problem is here and if we can modify the function/class to support both use cases somehow.

File: runner/src/indexer/indexer.ts [hunk 6]
```diff
@@ -463,13 +492,12 @@ export default class Indexer {
 
   // async writeLog (logEntry: LogEntry, logEntries: LogEntry[], functionName: string): Promise<any> {
```
Comment (darunrs) [inline runner/src/indexer/indexer.ts:493]: We don't need to pass in functionName here btw. We can sue the class variable this.indexerConfig to get the function name or full name or whatever it is you need.

File: runner/src/provisioner/provisioner.test.ts [hunk 7]
```diff
@@ -178,6 +180,18 @@ describe('Provisioner', () => {
       await expect(provisioner.provisionUserApi(accountId, functionName, databaseSchema)).rejects.toThrow('Failed to provision endpoint: Failed to add datasource: some error');
     });
 
+    it('throws an error when it fails to run sql to create indexer sql', async () => {
```
Comment (darunrs) [inline runner/src/provisioner/provisioner.test.ts:183]: nit: Use either run indexer sql or create indexer tables. Not a mix of both. Same for the below unit test. :)

**Review by darunrs — COMMENTED** [review]
Have you run this locally as well? Something like:
1. Create indexer with old code. 
2. Run indexer with new code and confirm logs table provisioned correctly
3. Create brand new indexer and confirm logs table provisioned correctly
4. Ran both of those indexers with logs and verified logs appeared in both

If so, can approve.

**Comment by Kevin101Zhang** [comment]
I ran locally in the following PR https://github.com/near/queryapi/pull/657. In the following PR the base is this current one.

**Review by darunrs — APPROVED** [review]
Nice, if it's all locally tested, then LGTM! Make sure to observe dev to confirm the provisioning works in dev. 

Check GCP logs for errors, check Hasura to see if indexers are running, and spot check particular indexers like access_keys_v1 to ensure that's running.

File: runner/src/indexer/indexer.ts [hunk 8]
```diff
@@ -160,7 +168,8 @@ export default class Indexer {
           } catch (e) {
             const error = e as Error;
             simultaneousPromises.push(this.writeLog(LogLevel.ERROR, functionName, blockHeight, 'Error running IndexerFunction', error.message));
-            // logEntries.push({ blockHeight, logTimestamp: new Date(), logType: LogType.SYSTEM, logLevel: LogLevel.ERROR, message: `Error running IndexerFunction ${error.message}` });
+            const indexerErrorLogEntry = LogEntry.systemError('Error running IndexerFunction', blockHeight);
```
Comment (Kevin101Zhang) [inline runner/src/indexer/indexer.ts:171]: forgot to comment this out as well

File: runner/src/provisioner/provisioner.test.ts (hunk 7 above)
Comment (Kevin101Zhang) [reply runner/src/provisioner/provisioner.test.ts:183]: Im a little confused - they are 2 separate test with 2 different error messages.
As oppose to calling provisioner.provisionUserApi im calling the methods directly and they both use executeSqlOnSchema to mock a reject error with different error messages.

provisioner.runIndexerSql
provisioner.runLogsSql

File: runner/src/indexer/indexer.ts (hunk 6 above)
Comment (Kevin101Zhang) [reply runner/src/indexer/indexer.ts:493]: Ah I see this [PR](https://github.com/near/queryapi/pull/646/files#diff-de4313f47fa61be86705f264cb2b269e977ff57b825b6c0987fc72ba4f35d8ea). Are you ok if I just add a ticket for this for when that PR is out to directly use `indexerConfig.functionName`?

**Pull request merged**